src/main/java/io/cdap/plugin/batch/source/MongoDBBatchSource.java

```diff
@@ -34,177 +33,201 @@
 
-  public MongoDBBatchSource(MongoDBConfig config) {
-    super(config);
+  public MongoDBBatchSource(MongoDBSourceConfig config) {
+    super(new ReferencePluginConfig(config.referenceName));
     this.config = config;
   }
 
   @Override
   public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
     super.configurePipeline(pipelineConfigurer);
+    config.validate();
     Schema schema = config.getSchema();
-    BSONConverter.validateSchema(schema);
     pipelineConfigurer.getStageConfigurer().setOutputSchema(schema);
   }
 
   @Override
   public void prepareRun(BatchSourceContext context) throws Exception {
+    config.validate();
     Configuration conf = new Configuration();
     conf.clear();
 
     MongoConfigUtil.setInputFormat(conf, MongoInputFormat.class);
-    MongoConfigUtil.setInputURI(conf, config.connectionString);
+    MongoConfigUtil.setInputURI(conf, config.getConnectionString());
+    MongoConfigUtil.setFields(conf, config.getProjectionDocument());
     if (!Strings.isNullOrEmpty(config.inputQuery)) {
       MongoConfigUtil.setQuery(conf, config.inputQuery);
     }
     if (!Strings.isNullOrEmpty(config.authConnectionString)) {
       MongoConfigUtil.setAuthURI(conf, config.authConnectionString);
     }
-    if (!Strings.isNullOrEmpty(config.inputFields)) {
-      MongoConfigUtil.setFields(conf, config.inputFields);
-    }
-    if (!Strings.isNullOrEmpty(config.splitterClass)) {
-      String className = String.format("%s.%s", StandaloneMongoSplitter.class.getPackage().getName(),
-                                       config.splitterClass);
-      Class<? extends MongoSplitter> klass = getClass().getClassLoader().loadClass(
-        className).asSubclass(MongoSplitter.class);
-      MongoConfigUtil.setSplitterClass(conf, klass);
-    }
-    LineageRecorder lineageRecorder = new LineageRecorder(context, config.referenceName);
-    lineageRecorder.createExternalDataset(config.getSchema());
+
+    emitLineage(context);
     context.setInput(Input.of(config.referenceName,
                               new SourceInputFormatProvider(MongoConfigUtil.getInputFormat(conf), conf)));
   }
 
   @Override
   public void initialize(BatchRuntimeContext context) throws Exception {
     super.initialize(context);
-    bsonConverter = new BSONConverter(config.getSchema());
+    bsonObjectToRecordTransformer = new BSONObjectToRecordTransformer(config.getSchema());
   }
 
   @Override
   public void transform(KeyValue<Object, BSONObject> input, Emitter<StructuredRecord> emitter) throws Exception {
+    try {
     BSONObject bsonObject = input.getValue();
-    emitter.emit(bsonConverter.transform(bsonObject));
+    emitter.emit(bsonObjectToRecordTransformer.transform(bsonObject));
+    } catch (Exception e) {
+      switch (config.getErrorHandling()) {
+        case SKIP:
+          LOG.warn("Failed to process record, skipping it", e);
+          break;
+        case FAIL_PIPELINE:
```

Comment:
there should be a 'send to error' option, that calls emitter.emitError()

Reply:
Option added.
```
    StructuredRecord transformed = null;
	    try {
	      transformed = bsonObjectToRecordTransformer.transform(input.getValue());
	      emitter.emit(transformed);
	    } catch (Exception e) {
	      switch (config.getErrorHandling()) {
	        case SEND_TO_ERROR:
	          emitter.emitError(new InvalidEntry<>(400, e.getMessage(), transformed));
	          break;
	
```
`emitError` expects an `InvalidEntry` instance with status and structured record set.
But there is no way to get structured record instance in this case. Also what status should be used?

Reply:
400 for the code is fine for now.

In this case we can use a StructuredRecord with just a single 'document' field, that contains a stringified version of the BSONObject. We also need to be sure to set the error schema in the configurePipeline() method using pipelineConfigurer.getStageConfigurer().setErrorSchema()

Reply:
Done.